Algorithm/QCAlgorithm.History.cs

```diff
@@ -624,34 +626,91 @@ public BaseData GetLastKnownPrice(Security security)
+        /// </remarks>
+        private IEnumerable<T> GetDataTypedHistory<T>(IEnumerable<HistoryRequest> requests, Symbol symbol)
             where T : IBaseData
         {
+            var type = typeof(T);
+
             if (requests == null || !requests.Any())
             {
                 throw new ArgumentException($"No history data could be fetched. " +
-                    $"This could be due to the specified security not being of the requested type. Symbol: {symbol} Requested Type: {typeof(T).Name}");
+                    $"This could be due to the specified security not being of the requested type. Symbol: {symbol} Requested Type: {type.Name}");
             }
 
             var slices = History(requests);
 
+            // If T is a custom data coming from Python (a class derived from PythonData), T will get here as PythonData
+            // and not the actual custom type. We take care of this especial case by using a dynamic version of GetDataTypedHistory that
+            // receives the Python type, and we get it from the history requests.
+            if (type == typeof(PythonData))
+            {
+                return GetPythonCustomDataTypeHistory(slices, requests.First().DataType, symbol).OfType<T>();
+            }
+
             // TODO: This is a patch to fix the issue with the Slice.GetImpl method returning only the last tick
             //       for each symbol instead of the whole list of ticks.
             //       The actual issue is Slice.GetImpl, so patch this can be removed right after it is properly addressed.
             //       A proposed solution making the Tick class a BaseDataCollection and make the Ticks class a dictionary Symbol->Tick instead of
             //       Symbol->List<Tick> so we can use the Slice.Get methods to collect all ticks in every slice instead of only the last one.
-            if (typeof(T) == typeof(Tick))
+            if (type == typeof(Tick))
             {
                 return (IEnumerable<T>)slices.Select(x => x.Ticks).Where(x => x.ContainsKey(symbol)).SelectMany(x => x[symbol]);
             }
 
-            return slices.Get<T>(symbol).Memoize();
+            return slices.Get<T>(symbol);
+        }
+
+        /// <summary>
+        /// Centralized logic to get data typed history for a given list of requests.
+        /// </summary>
+        /// <remarks>
+        /// This method will check for Python custom data types in order to call the right Slice.Get dynamic method
+        /// </remarks>
+        private IEnumerable<DataDictionary<T>> GetDataTypedHistory<T>(IEnumerable<HistoryRequest> requests)
+            where T : IBaseData
+        {
+            var slices = History(requests.Where(x => x != null));
+
+            if (typeof(T) == typeof(PythonData))
+            {
+                return GetPythonCustomDataTypeHistory(slices, requests.First().DataType).OfType<DataDictionary<T>>().Memoize();
+            }
+
+            return slices.Get<T>();
```

Comment:
- `History` is returning a `Memoize` already, we should probably use the private method which doesn't?
- This method has 2 returns, first using `Memoize`, second doesn't. Maybe can have a single return to avoid these issues?